backend/app/core/search.py

```diff
@@ -0,0 +1,466 @@
+            preference=self._opensearch_config.preference,
+        )
+        end = time.time()
+
+        # TODO: Log request time:
+        # f"query execution time: {round(end-start, 2)}s"
+
+        # TODO: Log response info:
+        # passage_hit_count = response['hits']['total']['value']
+        # # note: 'gte' values are returned when there are more than 10,000 results by default
+        # if response['hits']['total']['relation'] == "eq":
+        #     passage_hit_qualifier = "exactly"
+        # elif response['hits']['total']['relation'] == "gte":
+        #     passage_hit_qualifier = "at least"
+        # else:
+        #     passage_hit_qualifier = "unknown (unexpected)"
+        #
+        # doc_hit_count = response['aggregations']['sample']['bucketcount']['count']
+        # f"returned {passage_hit_qualifier} {passage_hit_count} passage(s) in {doc_hit_count} document(s)"
+
+        return OpenSearchResponse(
+            raw_response=response,
+            request_time_ms=round(1000 * (end - start)),
+        )
+
+
+def _innerproduct_threshold_to_lucene_threshold(ip_thresh: float) -> float:
+    """Maps inner product to lucene threashold.
+
+    Opensearch documentation on mapping similarity functions to Lucene thresholds is
+    here: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/SpaceType.java#L33
+
+    It defines 'inner product' as negative inner product i.e. a distance rather than
+    similarity measure, so we reverse the signs of inner product here compared to the docs.
+    """
+    if ip_thresh > 0:
+        return ip_thresh + 1
+    else:
+        return 1 / (1 - ip_thresh)
+
+
+def _year_range_filter(
+    year_range: Tuple[Optional[int], Optional[int]]
+) -> Optional[Dict[str, Any]]:
+    """Get an Opensearch filter for year range.
+
+    The filter returned is between the first term of `year_range` and the last term,
+    and is inclusive. Either value can be set to None to only apply one year constraint.
+    """
+
+    policy_year_conditions = {}
+    if year_range[0] is not None:
+        policy_year_conditions["gte"] = f"01/01/{year_range[0]}"
+    if year_range[1] is not None:
+        policy_year_conditions["lte"] = f"31/12/{year_range[1]}"
+
+    if policy_year_conditions:
+        return {"range": {"action_date": policy_year_conditions}}
+
+    return None
+
+
+@dataclass(frozen=True)
+class SearchConfig:
+    """Configuration for searches sent to OpenSearch."""
+
+    name_boost: int = _DEFAULT_NAME_BOOST
+    description_boost: int = _DEFAULT_DESCRIPTION_BOOST
+    lucene_threshold: float = _innerproduct_threshold_to_lucene_threshold(
+        _DEFAULT_INNER_PRODUCT_THRESHOLD
+    )  # TODO: tune me separately for descriptions?
+    max_doc_count: int = _DEFAULT_MAX_DOC_COUNT
+    max_passages_per_doc: int = _DEFAULT_MAX_PASSAGES_PER_DOC
+    n_passages_to_sample_per_shard = _DEFAULT_N_PASSAGES_TO_SAMPLE_PER_SHARD
+    k = 10000  # TODO: tune me
```

Comment:
Could we assign this at the top as with `max_passages_per_doc` etc, as its value will probably need to be tweaked in the same thought process as some of the others? Maybe `_DEFAULT_KNN_K_VALUE`?